File: all/1722_minimize_hamming_distance_after_swap_operations/sol_better_iterative_find.py
from collections import Counter
from typing import List


class UnionFindDS:
    def __init__(self, n: int):
        self._parents_list = [i for i in range(n)]

    def find(self, i: int) -> int:
        if i >= (len(self._parents_list)):
            return -1

        root = i

        while root != ((self._parents_list)[root]):
            root = ((self._parents_list)[root])

        # Do path compresion to make all the nodes that have the same ancestor
        # point to the ancestor directly.
        while i != ((self._parents_list)[i]):
            original_parent = ((self._parents_list)[i])
            self._parents_list[i] = root
            i = original_parent

        return root

    def union(self, i: int, j: int) -> None:
        root_of_i = self.find(i)
        root_of_j = self.find(j)

        if root_of_i == root_of_j:
            # Return if i and j are in the same group.
            return

        # i and j are not in the same group, make the root of the root of j
        # become i, so i and j can be in the same group.
        self._parents_list[root_of_j] = root_of_i


class Solution:
    def minimumHammingDistance(
        self,
        source: List[int],
        target: List[int],
        allowedSwaps: List[List[int]]
    ) -> int:
        minimum_hamming_distance = 0
        union_find_ds_instance = UnionFindDS(len(source))

        # Set union find data structure.
        for i, j in allowedSwaps:
            union_find_ds_instance.union(i, j)

        # For each group of indices, we need count what numbers are in each
        # group and their counts.
        group_and_numbers_counter_mapping_dict = {}

        for i in range(len(source)):
            root_of_i = union_find_ds_instance.find(i)

            if root_of_i not in group_and_numbers_counter_mapping_dict:
                group_and_numbers_counter_mapping_dict[root_of_i] = Counter()

            group_and_numbers_counter_mapping_dict[root_of_i][source[i]] += 1

        # Iterate the target list, check if current iterated number can be
        # fulfilled with a certain group of numbers in the source list.
        for index, value in enumerate(target):
            # Get the group of current iterated index.
            root_of_index = union_find_ds_instance.find(index)

            if not group_and_numbers_counter_mapping_dict[root_of_index][value]:
                # Increase minimum hamming distance by if current iterated
                # number cannot be fulfilled by current group of numbers.
                minimum_hamming_distance += 1
            else:
                # Decrease the corresponding number counter by 1 as a sign of
                # fulfillment.
                group_and_numbers_counter_mapping_dict[root_of_index][value] -= (
                    1
                )

        return minimum_hamming_distance


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        [[1, 2, 3, 4], [2, 1, 4, 5], [[0, 1], [2, 3]]],
        [[1, 2, 3, 4], [1, 3, 2, 4], []],
        [[5, 1, 2, 4, 3], [1, 5, 4, 2, 3], [[0, 4], [4, 2], [1, 3], [1, 4]]],
    ]

    for testcase in testcases_list:
        print(sol.minimumHammingDistance(*testcase))

File: all/1722_minimize_hamming_distance_after_swap_operations/sol_better_iterative_find.c
#include <stddef.h>
#include <stdlib.h>

typedef struct {
    int root_value;
    int value;
} RootValueAndValuePairST;

static int RootValueAndValuePairSTCompareFunction(
    const void *p1,
    const void *p2
) {
    const RootValueAndValuePairST *rvp_ptr1 = (
        (const RootValueAndValuePairST*)p1
    );
    const RootValueAndValuePairST *rvp_ptr2 = (
        (const RootValueAndValuePairST*)p2
    );

    if ((rvp_ptr1->root_value) == (rvp_ptr2->root_value)) {
        return (
            ((rvp_ptr1->value) > (rvp_ptr2->value))
            -
            ((rvp_ptr1->value) < (rvp_ptr2->value))
        );
    }

    return (
        ((rvp_ptr1->root_value) > (rvp_ptr2->root_value))
        -
        ((rvp_ptr1->root_value) < (rvp_ptr2->root_value))
    );
}

static int unionFindFind(
    int *parents_array_ptr,
    const int parents_array_length,
    const int i
) {
    if (i >= parents_array_length) {
        return -1;
    }

    int root = i, original_parent = 0;
    int current_i = i;

    while (root != (parents_array_ptr[root])) {
        root = (parents_array_ptr[root]);
    }

    /*
     * Do path compresion to make all the nodes that have same ancestor point to
     * the ancestor directly.
     */
    while (root != (parents_array_ptr[current_i])) {
        original_parent = (parents_array_ptr[current_i]);
        parents_array_ptr[current_i] = root;
        current_i = original_parent;
    }

    return root;
}

static void unionFindUnion(
    int * parents_array_ptr,
    const int parents_array_length,
    const int i,
    const int j
) {
    const int root_of_i = unionFindFind(
        parents_array_ptr,
        parents_array_length,
        i
    );
    const int root_of_j = unionFindFind(
        parents_array_ptr,
        parents_array_length,
        j
    );

    if (root_of_i == root_of_j) {
        return;
    }

    parents_array_ptr[root_of_j] = root_of_i;
}

int minimumHammingDistance(
    int* source,
    int sourceSize,
    int* target,
    int targetSize,
    int** allowedSwaps,
    int allowedSwapsSize,
    int* allowedSwapsColSize
) {
    int minimum_hamming_distance = sourceSize;
    int *parents_array_ptr = (
        (int*)malloc(sizeof(*parents_array_ptr) * sourceSize)
    );
    RootValueAndValuePairST *root_value_and_value_pairs_st_array_ptr = (
        (RootValueAndValuePairST*)malloc(
            sizeof(*root_value_and_value_pairs_st_array_ptr) * sourceSize * 2
        )
    );
    RootValueAndValuePairST *source_root_value_and_value_pairs_st_array_ptr = (
        root_value_and_value_pairs_st_array_ptr
    );
    RootValueAndValuePairST *target_root_value_and_value_pairs_st_array_ptr = (
        root_value_and_value_pairs_st_array_ptr + sourceSize
    );

    for (int i = 0; i < sourceSize; ++i) {
        parents_array_ptr[i] = i;
    }

    /* Set union find data structure. */
    for (int k = 0, i = 0, j = 0; k < allowedSwapsSize; ++k) {
        i = (allowedSwaps[k][0]);
        j = (allowedSwaps[k][1]);

        unionFindUnion(parents_array_ptr, sourceSize, i, j);
    }

    /*
     * Construct the root value (parent value) and value pairs array for source
     * array and target array.
     */
    for (int i = 0, root_of_i = 0; i < sourceSize; ++i) {
        root_of_i = unionFindFind(parents_array_ptr, sourceSize, i);

        (source_root_value_and_value_pairs_st_array_ptr[i]).root_value = (
            root_of_i
        );
        (source_root_value_and_value_pairs_st_array_ptr[i]).value = (source[i]);
        (target_root_value_and_value_pairs_st_array_ptr[i]).root_value = (
            root_of_i
        );
        (target_root_value_and_value_pairs_st_array_ptr[i]).value = (target[i]);
    }

    /*
     * Sort the source root value and value pairs array.
     * After sorting, the indices that are in the same group will be grouped
     * together and their values will be sorted too.
     */
    qsort(
        (void*)source_root_value_and_value_pairs_st_array_ptr,
        sourceSize,
        sizeof(*source_root_value_and_value_pairs_st_array_ptr),
        &RootValueAndValuePairSTCompareFunction
    );

    /*
     * Sort the target root value and value pairs array.
     * After sorting, the indices that are in the same group will be grouped
     * together and their values will be sorted too.
     */
    qsort(
        (void*)target_root_value_and_value_pairs_st_array_ptr,
        targetSize,
        sizeof(*target_root_value_and_value_pairs_st_array_ptr),
        &RootValueAndValuePairSTCompareFunction
    );

    /*
     * Use two pointers to iterated the source root value and value pairs
     * array and the target root value and value pairs array to find how
     * many numbers in the target array can be fulfilled by swapping some
     * numbers in the source array under the given swap pairs constraint.
     */
    for (int i = 0, j = 0; (i < sourceSize) && (j < targetSize);) {
        if (
            ((source_root_value_and_value_pairs_st_array_ptr[i]).root_value)
            ==
            ((target_root_value_and_value_pairs_st_array_ptr[j]).root_value)
        ) {
            if (
                ((source_root_value_and_value_pairs_st_array_ptr[i]).value)
                ==
                ((target_root_value_and_value_pairs_st_array_ptr[j]).value)
            ) {
                --minimum_hamming_distance;
                ++i;
                ++j;
            }
            else if (
                ((source_root_value_and_value_pairs_st_array_ptr[i]).value)
                <
                ((target_root_value_and_value_pairs_st_array_ptr[j]).value)
            ) {
                ++i;
            }
            else {
                ++j;
            }
        }
        else if (i > j) {
            j = i;
        }
        else {
            i = j;
        }
    }

    free(parents_array_ptr);
    parents_array_ptr = NULL;

    free(root_value_and_value_pairs_st_array_ptr);
    root_value_and_value_pairs_st_array_ptr = NULL;
    source_root_value_and_value_pairs_st_array_ptr = NULL;
    target_root_value_and_value_pairs_st_array_ptr = NULL;

    return minimum_hamming_distance;
}
